homeassistant/components/qbus/coordinator.py

```diff
@@ -0,0 +1,204 @@
+"""Qbus coordinator."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+
+from qbusmqttapi.discovery import QbusDiscovery, QbusMqttDevice
+from qbusmqttapi.factory import QbusMqttMessageFactory, QbusMqttTopicFactory
+
+from homeassistant.components.mqtt import ReceiveMessage, client as mqtt
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_call_later
+
+from .const import CONF_SERIAL
+from .entity import QbusEntity
+from .qbus import QbusConfigContainer
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class QbusRuntimeData:
+    """Runtime data for Qbus integration."""
+
+    coordinator: QbusDataCoordinator
+
+
+type QbusConfigEntry = ConfigEntry[QbusRuntimeData]
+
+
+class QbusDataCoordinator:
+    """Qbus data coordinator."""
+
+    _WAIT_TIME = 3
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize Qbus coordinator."""
+        _LOGGER.debug("Initializing coordinator %s", entry.entry_id)
+
+        self._message_factory = QbusMqttMessageFactory()
+        self._topic_factory = QbusMqttTopicFactory()
+
+        self._hass = hass
+        self._entry = entry
+
+        self._platform_register: dict[
+            str, tuple[type[QbusEntity], AddEntitiesCallback]
+        ] = {}
+        self._registered_entity_ids: list[str] = []
+        self._device_activated = False
+        self._subscribed_to_device_state = False
+        self._device: QbusMqttDevice | None
+
+        # Clean up when HA stops
+        entry.async_on_unload(
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
+        )
+
+        # Subscribe to option updates
+        entry.async_on_unload(entry.add_update_listener(self._options_update_listener))
+
+    def shutdown(self, event: Event | None = None) -> None:
+        """Shutdown Qbus coordinator."""
+        _LOGGER.debug("Shutting down coordinator for entry %s", self._entry.entry_id)
+
+        self._registered_entity_ids = []
+        self._device_activated = False
+        self._subscribed_to_device_state = False
+
+    def remove(self) -> None:
```

Comment:
We can remove this method and call shutdown() directly

Reply:
Right!